pkg/sql/pai_submitter.go

```diff
@@ -236,85 +239,137 @@ func (s *paiSubmitter) ExecuteExplain(cl *ir.ExplainStmt) error {
 		if err != nil {
 			return err
 		}
-		err = createExplainResultTable(db, cl, cl.Into, isDeepModel)
+		err = createExplainResultTable(db, cl, cl.Into, modelType, estimator)
 		if err != nil {
 			return err
 		}
 	}
 	scriptPath := fmt.Sprintf("file://%s/%s", s.Cwd, tarball)
-	code, paiCmd, requirements, e := pai.Explain(cl, s.Session, scriptPath, cl.ModelName, ossModelPath, s.Cwd, isDeepModel)
+	code, paiCmd, requirements, e := pai.Explain(cl, s.Session, scriptPath, cl.ModelName, ossModelPath, s.Cwd, modelType)
 	if e != nil {
 		return e
 	}
 	return s.submitPAITask(code, paiCmd, requirements)
 }
 
-func ossModelFileExists(modelName string) (bool, error) {
+// getOSSSavedModelType returns the saved model type when training, can be:
+// 1. randomforests: model is saved by pai
+// 2. xgboost: on OSS with model file xgboost_model_desx
+// 3. PAI tensorflow models: on OSS with meta file: tensorflow_model_desc
+func getOSSSavedModelType(modelName string) (modelType int, estimator string, err error) {
 	// FIXME(typhoonzero): if the model not exist on OSS, assume it's a random forest model
 	// should use a general method to fetch the model and see the model type.
 	endpoint := os.Getenv("SQLFLOW_OSS_MODEL_ENDPOINT")
 	ak := os.Getenv("SQLFLOW_OSS_AK")
 	sk := os.Getenv("SQLFLOW_OSS_SK")
 	if endpoint == "" || ak == "" || sk == "" {
-		return false, fmt.Errorf("must define SQLFLOW_OSS_MODEL_ENDPOINT, SQLFLOW_OSS_AK, SQLFLOW_OSS_SK when using submitter maxcompute")
+		err = fmt.Errorf("must define SQLFLOW_OSS_MODEL_ENDPOINT, SQLFLOW_OSS_AK, SQLFLOW_OSS_SK when using submitter maxcompute")
+		return
 	}
 	// NOTE(typhoonzero): PAI Tensorflow need SQLFLOW_OSS_CHECKPOINT_DIR, get bucket name from it
 	ossCheckpointDir := os.Getenv("SQLFLOW_OSS_CHECKPOINT_DIR")
 	ckptParts := strings.Split(ossCheckpointDir, "?")
 	if len(ckptParts) != 2 {
-		return false, fmt.Errorf("SQLFLOW_OSS_CHECKPOINT_DIR got wrong format")
+		err = fmt.Errorf("SQLFLOW_OSS_CHECKPOINT_DIR got wrong format")
+		return
 	}
 	urlParts := strings.Split(ckptParts[0], "://")
 	if len(urlParts) != 2 {
-		return false, fmt.Errorf("SQLFLOW_OSS_CHECKPOINT_DIR got wrong format")
+		err = fmt.Errorf("SQLFLOW_OSS_CHECKPOINT_DIR got wrong format")
 	}
 	bucketName := strings.Split(urlParts[1], "/")[0]
 
 	cli, err := oss.New(endpoint, ak, sk)
 	if err != nil {
-		return false, err
+		return
 	}
 	bucket, err := cli.Bucket(bucketName)
 	if err != nil {
-		return false, err
+		return
+	}
+	ret, err := bucket.IsObjectExist(modelName + "/tensorflow_model_desc")
+	if err != nil {
+		return
+	}
+	if ret {
+		modelType = pai.ModelTypeTF
+		var buf []byte
+		err = bucket.GetObjectToFile(modelName+"/tensorflow_model_desc_estimator", "tmp_estimator_name")
```

Comment:
Maybe we can use `GetObject` to avoid create a tmp file:
``` golang
body, err := bucket.GetObject(objectKey)
	if err != nil {
		HandleError(err)
	}

	data, err := ioutil.ReadAll(body)
	body.Close()
```
Ref: https://github.com/aliyun/aliyun-oss-go-sdk/blob/master/sample/get_object.go#L28

Reply:
I tried it without success and switched back to `GetObjectToFile`.